### PR TITLE
Add possibility to extract tech from import declaration.

### LINKED ISSUE #11: Support custom extension for entity techs
It seems that it only supports js modules. How to use it with TypeScript?

https://github.com/bem/webpack-bem-loader/blob/master/index.js#L10

### LINKED ISSUE #14: Wrong entity for block_mod inside block__elem
Example:
_page__content.react.js_
```js
import 'b:page m:theme=normal';
...
```
--> { block: 'b-page', elem: 'content', modName: 'theme', modVal: 'normal' }

### PR DESCRIPTION
**Pull request by Yeti-or**
Close #14 
Close #11 
@awinogradov take a look

### CODE REVIEW

**Comment by Yeti-or** [comment]
Checked on `bem/bem-react-core` `bem/bem-react-components` and `islands`

File: index.js [hunk 1]
```diff
@@ -2,37 +2,29 @@
...
     vowFs = require('vow-fs'),
-    loaderUtils = require('loader-utils'),
-    defaultNaming = { elem : '-', elemDirPrefix: '', modDirPrefix: '_' },
     isFileJsModule = file => path.extname(file) === '.js';
 
 module.exports = function(source) {
     this.cacheable && this.cacheable();
 
     const callback = this.async(),
-        options = this.options.bemLoader || loaderUtils.parseQuery(this.query),
```
Comment (awinogradov) [inline index.js]: Вообще-то это поддержка webpack2, верни это на место и строку 8 тоже)

Comment (Yeti-or) [inline index.js]: Плохо отребейзился походу, ща верну

**Comment by veged** [comment]
there is a problem with our approach with `import MyBlockI18N from 'b:MyBlock.i18n';` — for elems and mods syntax gonna be awkward (like `import MyElemI18N from 'b:MyBlock e:MyElem.i18n';`)

maybe it's better to stick to initial idea with importing "magic" `i18n` module with accessors to all keysets

let's discuss this offline with @awinogradov

**Comment by awinogradov** [comment]
@veged, I suggest `import B from 'b:B e:E t:i18n'`. It'll help us to support CSS-only import `import B from 'b:B t:css'`

**Comment by veged** [comment]
@awinogradov why we need CSS-only import?

**Comment by awinogradov** [comment]
@veged you can have block with js and css and you may want to require css only. it can be block with animations or smth else. Also, I think this
``` js
import B from 'b:B e:E t:i18n';
import B from 'b:B t:css'
```
looks like deps by technology and looks better for block and element import:
``` js
// block.deps.js
[{
  tech: 'js',
  shouldDeps: {
    block: 'b'
},
{
  tech: 'bemhtml',
  shouldDeps: {
    block: 'b'
   }
}]
```

File: index.js [hunk 2]
```diff
@@ -15,77 +18,122 @@ module.exports = function(source) {
...
-                            n = `[${requires.join(',')}]${idx? `[${requireIdx}]` : ''}`;
-
-                        node.update(idx? `
-                          (${n}.default?
-                              ${n}.default.applyDecls() :
-                              ${n}.applyDecls())
-                          ` : n);
-                    }));
+                    });
+                    if (arr.length - 1 === i) {
```
Comment (veged) [inline index.js]: seems like unnecessary

File: generators/index.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * @param {BemFile[]} files
+ * @returns {String}
+ */
+function generateReactStr(files) {
+    const apply = require => `${require}.default ?
+        ${require}.default.applyDecls() :
+        ${require}.applyDecls()`;
+
+    const reqStr = file => {
```
Comment (Yeti-or) [inline generators/index.js]: `(${require}.default || ${require}).applyDecls()`

File: generators/index.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    return files
+        .reduce((acc, file, i) => acc.concat(
+            i !== files.length - 1
+                ? `require('${file.path}'),`
+                : `(require('${file.path}').default || require('${file.path}')).applyDecls()`
+        ), ['('])
+        .concat(');')
+        .join('\n');
+}
+
```
Comment (awinogradov) [inline generators/index.js]: Unnecessary line

File: index.js [hunk 5]
```diff
@@ -1,142 +1,142 @@
...
+        fastTechMap = Object.keys(techMap).reduce((acc, tech) => {
+            techMap[tech].forEach(_tech => {
+                acc[_tech] = tech;
+            });
+            return acc;
+        }, {}),
+        defaultTechs = Object.keys(techMap).reduce((acc, tech) => {
+            return acc.concat(techMap[tech])
+        }, []),
+        generators = Object.assign(require('./generators'), options.customGenerators),
```
Comment (awinogradov) [inline index.js]: ?

File: index.js [hunk 6]
```diff
@@ -1,142 +1,142 @@
...
+                    const entityPath = path.resolve(
+                            process.cwd(),
+                            bemFs.path(bemCell, namingOptions)
+                        );
+
+                    this.addDependency(entityPath);
+
+                    return vowFs
+                        .exists(entityPath)
+                        .then(exist => {
```
Comment (awinogradov) [inline index.js]: Can be written in one line

File: index.js [hunk 7]
```diff
@@ -1,142 +1,142 @@
...
+                                    ) {
+                                        messages.map(this.emitError, this);
+                                    }
+                                }
+                            });
+
+                            // Each tech has own generator
+                            const value = Object.keys(techToEntities).map(tech => {
+                                // _tech - is tech from options.techMap
+                                // but it could be processed with customGenerator so we need to check just tech too.
```
Comment (awinogradov) [inline index.js]: Also about generators?

File: index.js [hunk 8]
```diff
@@ -1,142 +1,142 @@
...
+
+                            node.update(value);
+                        })
+                );
             }
-        });
+    });
 
     vow.all(allPromises)
         .then(() => {
```
Comment (awinogradov) [inline index.js]: One line?

File: index.js [hunk 9]
```diff
@@ -1,142 +1,142 @@
...
+                )
+
+                // expand entities by all provided levels
+                .reduce((acc, entity) => {
+                    levels.forEach(layer => {
+                        // if entity has tech get exactly it,
+                        // otherwise expand entities by default techs
+                        [].concat(entity.tech || defaultTechs).forEach(tech => {
+                            const cell = new BemCell({entity: new BemEntityName(entity), tech, layer});
+                            acc.push(cell);
```
Comment (dfilatov) [inline index.js]: remove redundant `const`

File: index.js [hunk 10]
```diff
@@ -1,142 +1,142 @@
...
+                            const techToEntities = {};
+                            /**
+                             * techToEntities:
+                             *   js: [enity, entity]
+                             *   css: [entity, entity, entity]
+                             *   i18n: [entity]
+                             */
+                            bemFiles.forEach(file => {
+                                if (file.exist) {
+                                    techToEntities[file.cell.tech] = (techToEntities[file.cell.tech] || []).concat(file);
```
Comment (dfilatov) [inline index.js]: `(obj[k] || (obj[k] = [])).push(x)`

File: index.js [hunk 11]
```diff
@@ -1,142 +1,142 @@
...
+
+                            // Each tech has own generator
+                            const value = Object.keys(techToEntities).map(tech => {
+                                // _tech - is tech from options.techMap
+                                // but it could be processed with customGenerator so we need to check just tech too.
+                                const _tech = fastTechMap[tech] || tech;
+                                return generators[_tech]?
+                                    generators[_tech](techToEntities[tech]) :
+                                    generators['*'](techToEntities[tech]);
+                            }).join('\n')
```
Comment (dfilatov) [inline index.js]:
```js
(generators[_tech] || generators['*'])(techToEntities[tech])
```

File: index.js [hunk 12]
```diff
@@ -1,142 +1,142 @@
...
+                                // check if entity has no tech to resolve
+                                if (possibleErrors[fileId].length === defaultTechs.length) {
+                                    const messages = [];
+                                    if (
+                                        possibleErrors[fileId]
+                                            .every(file => {
+                                                messages.push(`BEM-Module not found: ${file.path}`);
+                                                return ~defaultTechs.indexOf(file.cell.tech);
+                                            })
+                                    ) {
```
Comment (dfilatov) [inline index.js]: &&

File: index.js (hunk 10 above)
Comment (Yeti-or) [inline index.js]: we need something to require for each bemEntity

File: index.js (hunk 8 above)
Comment (Yeti-or) [inline index.js]: you mean `then(() => callback(null, result.toString()))`

File: index.js [hunk 13]
```diff
@@ -1,142 +1,126 @@
...
+            return acc;
+        }, {}),
+        extToTech = Object.keys(techMap).reduce((acc, tech) => {
+            techMap[tech].forEach(_tech => {
+                acc[_tech] = tech;
+            });
+            return acc;
+        }, {}),
+        defaultTechs = Object.keys(techMap).reduce((acc, tech) => acc.concat(techMap[tech]), []),
+        generators = Object.assign(require('./generators'), options.customGenerators),
```
Comment (awinogradov) [inline index.js]: customGenerators?

File: index.js [hunk 14]
```diff
@@ -88,34 +89,36 @@ module.exports = function(source) {
...
                             bemFiles.forEach(file => {
-                                existsEntities[file.cell.entity.id] || (existsEntities[file.cell.entity.id] = []);
-                                unExistsEntities[file.cell.entity.id] || (unExistsEntities[file.cell.entity.id] = []);
-                                uniqEntities[file.cell.entity.id] = file.cell.entity;
                                 if (file.exist) {
                                     (techToFiles[file.cell.tech] || (techToFiles[file.cell.tech] = [])).push(file);
-                                    existsEntities[file.cell.entity.id].push(file);
+                                    existsEntities[file.cell.entity.id] = true;
+
+                                    if (file.cell.entity.mod && !file.cell.entity.isSimpleMod()) {
```
Comment (zxqfox) [inline index.js]: можно же было геттер сделать o_O

File: package.json [hunk 15]
```diff
@@ -12,7 +12,8 @@
   ],
   "license": "MPL-2.0",
   "dependencies": {
-    "@bem/cell": "^0.2.4",
+    "@bem/cell": "^0.2.2",
```
Comment (zxqfox) [inline package.json]: зачем 0.2.4→0.2.2?

File: index.js (hunk 14 above)
Comment (Yeti-or) [inline index.js]: хм, не знаю как так вышло

File: package.json (hunk 15 above)
Comment (Yeti-or) [inline package.json]: хм, не знаю как так вышло

File: index.js (hunk 14 above)
Comment (Yeti-or) [inline index.js]: не понял тебя

File: index.js [hunk 16]
```diff
@@ -1,142 +1,141 @@
...
+                            entity.isSimpleMod() === false &&
+                                (existsEntities[BemEntityName.create({ block, elem, modName }).id] = true);
+                        });
 
-                        const idx = requireIdx !== null,
-                            n = `[${requires.join(',')}]${idx? `[${requireIdx}]` : ''}`;
+                        Object.keys(existsEntities).forEach(fileId => {
+                            // check if entity has no tech to resolve
+                            existsEntities[fileId] || errEntities[fileId].forEach(file => {
+                                this.emitError(`BEM-Module not found: ${file.path}`);
```
Comment (Yeti-or) [inline index.js]: `BEM module not found: ${file.path}`

Comment (Yeti-or) [inline index.js]: Агрегировать в одну ошибку для одной сущности

Comment (Yeti-or) [inline index.js]: `{common,desktop}/button2.{css,js,stylus}`

**Pull request merged**